### PR TITLE
check for existing transaction created with visa rtf

### PR DESCRIPTION
**Pull request by NileshPant1999**

### CODE REVIEW

File: apps/travelperk/helpers.py [hunk 1]
```diff
@@ -112,29 +113,68 @@ def get_employee_email(platform_connection, employee_email):
...
+        'amount': f'eq.{amount}',
+        #'spent_at': f'eq.{str(datetime.strptime(expense.expense_date, "%Y-%m-%d").replace(tzinfo=timezone.utc))}',
+        'offset': 0,
+        'limit': 1
+    }
+
+    # Fetch expense details from Fyle
+    expense_details = platform_connection.v1beta.admin.corporate_card_transactions.list(query_params)['data']
+
+    # Return True if any expense details are found, False otherwise
```
Comment (Shwetabhk) [inline apps/travelperk/helpers.py:142]: The comment seems wrong we're returning an array of dictionary right?
Doesn't seem like we're returning True or False

File: apps/travelperk/helpers.py [hunk 2]
```diff
@@ -112,29 +113,68 @@ def get_employee_email(platform_connection, employee_email):
...
+    """
+    Check for Duplicate Transaction in Fyle
+
+    Args:
+    - platform_connection: Fyle platform connection object.
+    - expense: Expense object containing details of the transaction.
+    - corporate_card_id: ID of the corporate card used for the transaction.
+
+    Returns:
+    - True if a duplicate transaction is found, False otherwise.
```
Comment (Shwetabhk) [inline apps/travelperk/helpers.py:126]: we're not returning true or false

File: apps/travelperk/helpers.py [hunk 3]
```diff
@@ -112,29 +113,68 @@ def get_employee_email(platform_connection, employee_email):
...
+        'order': 'updated_at.asc',
+        'corporate_card_id': f'eq.{corporate_card_id}',
+        'amount': f'eq.{amount}',
+        #'spent_at': f'eq.{str(datetime.strptime(expense.expense_date, "%Y-%m-%d").replace(tzinfo=timezone.utc))}',
+        'offset': 0,
+        'limit': 1
+    }
+
+    # Fetch expense details from Fyle
+    expense_details = platform_connection.v1beta.admin.corporate_card_transactions.list(query_params)['data']
```
Comment (Shwetabhk) [inline apps/travelperk/helpers.py:140]: Please test well for case where 0 Transactions are returned. 
We're referencing `data` so ensure `data` key is present even if 0 results are returned

File: apps/travelperk/actions.py [hunk 4]
```diff
@@ -204,42 +211,42 @@ def create_invoice_lineitems(org_id, invoice, expense, user_role, amount):
...
     platform_connection = create_fyle_connection(org_id)
+    matched_transaction_found = []
 
+    # Determine the employee's email based on the user role
     if user_role in ['TRAVELLER', 'BOOKER']:
         travelperk_employee = getattr(expense, ROLE_EMAIL_MAPPING.get(user_role, None))
         employee_email = get_employee_email(platform_connection, travelperk_employee)
     else:
-        employee_email = get_email_from_credit_card(platform_connection, expense.credit_card_last_4_digits)
+        employee_email, matched_transaction_found = get_email_from_credit_card(platform_connection, expense, amount)
```
Comment (Shwetabhk) [inline apps/travelperk/actions.py:223]: why are we doing this in 1 function? the function name doesn't match the action we're doing
we're getting emails as well as matched transaction

File: apps/travelperk/actions.py [hunk 5]
```diff
@@ -204,42 +211,42 @@ def create_invoice_lineitems(org_id, invoice, expense, user_role, amount):
         object: Created expense object.
     """
 
-    # Determine the employee's email based on the user role
+    # Establish a connection to the Fyle platform
     platform_connection = create_fyle_connection(org_id)
+    matched_transaction_found = []
```
Comment (Shwetabhk) [inline apps/travelperk/actions.py:216]:
```suggestion
    matched_transaction = []
```

File: apps/travelperk/actions.py [hunk 6]
```diff
@@ -204,42 +211,42 @@ def create_invoice_lineitems(org_id, invoice, expense, user_role, amount):
...
         employee_email = get_employee_email(platform_connection, travelperk_employee)
     else:
-        employee_email = get_email_from_credit_card(platform_connection, expense.credit_card_last_4_digits)
+        employee_email, matched_transaction_found = get_email_from_credit_card(platform_connection, expense, amount)
 
-    # Create the payload for the expense
-    payload = construct_expense_payload(org_id, expense, amount)
+    file_ids = construct_file_ids(platform_connection, invoice.pdf)
+
+    if not len(matched_transaction_found):
```
Comment (Shwetabhk) [inline apps/travelperk/actions.py:227]:
```suggestion
    if not matched_transaction_found:
```
len() is not required

File: apps/travelperk/helpers.py [hunk 7]
```diff
@@ -112,29 +113,68 @@ def get_employee_email(platform_connection, employee_email):
...
+    Returns:
+    - True if a duplicate transaction is found, False otherwise.
+    """
+
+    # Prepare query parameters
+    query_params = {
+        'order': 'updated_at.asc',
+        'corporate_card_id': f'eq.{corporate_card_id}',
+        'amount': f'eq.{amount}',
+        #'spent_at': f'eq.{str(datetime.strptime(expense.expense_date, "%Y-%m-%d").replace(tzinfo=timezone.utc))}',
```
Comment (Shwetabhk) [inline apps/travelperk/helpers.py:134]: date part seems to be commented, how are we ensuring that the spend date matches?

File: apps/travelperk/helpers.py [hunk 8]
```diff
@@ -112,29 +113,68 @@ def get_employee_email(platform_connection, employee_email):
...
+    }
+
+    # Fetch expense details from Fyle
+    expense_details = platform_connection.v1beta.admin.corporate_card_transactions.list(query_params)['data']
+
+    # Return True if any expense details are found, False otherwise
+    return expense_details
+
+
+def get_email_from_credit_card(platform_connection, expense, amount):
```
Comment (Shwetabhk) [inline apps/travelperk/helpers.py:146]: function is doing 2 actions, name doesn't represent that
better to call both function separately

Comment (Shwetabhk) [reply apps/travelperk/helpers.py:146]: don't see any reason to call them together

Comment (Shwetabhk) [reply apps/travelperk/helpers.py:146]: If you need to call the together for figuring out the email, then please name the function appropriately

File: apps/travelperk/helpers.py [hunk 9]
```diff
@@ -143,6 +183,30 @@ def get_email_from_credit_card(platform_connection, credit_card_last_4_digits):
...
 
         employee = platform_connection.v1beta.admin.employees.list(query_params)['data']
-        return employee[0]['user']['email'] if employee else None
 
-    return None
+        # Return user's email and matched transaction status
+        return (employee[0]['user']['email'], matched_transaction_found)
+
+    # Return None and matched transaction status if user ID is not found or a matched transaction is found
+    return (None, matched_transaction_found)
```
Comment (Shwetabhk) [inline apps/travelperk/helpers.py:191]: parenthesis is not required

File: apps/travelperk/helpers.py [hunk 10]
```diff
@@ -143,6 +183,30 @@ def get_email_from_credit_card(platform_connection, credit_card_last_4_digits):
         }
 
         employee = platform_connection.v1beta.admin.employees.list(query_params)['data']
-        return employee[0]['user']['email'] if employee else None
 
-    return None
+        # Return user's email and matched transaction status
+        return (employee[0]['user']['email'], matched_transaction_found)
```
Comment (Shwetabhk) [inline apps/travelperk/helpers.py:188]: parenthesis not required

File: apps/travelperk/helpers.py [hunk 11]
```diff
@@ -112,29 +113,68 @@ def get_employee_email(platform_connection, employee_email):
...
 
     credit_card_details = platform_connection.v1beta.admin.corporate_cards.list(query_params)['data']
-
+    # Retrieve user ID and corporate credit card ID
     user_id = credit_card_details[0]['user_id'] if credit_card_details else None
+    corporate_credit_card_id = credit_card_details[0]['id'] if credit_card_details else None
 
-    if user_id:
+    # Check for matched transaction
+    matched_transaction_found = check_for_transaction_in_fyle(platform_connection, expense, corporate_credit_card_id, amount)
```
Comment (Shwetabhk) [inline apps/travelperk/helpers.py:172]:
```suggestion
    matched_transaction = check_for_transaction_in_fyle(platform_connection, expense, corporate_credit_card_id, amount)
```

File: apps/travelperk/helpers.py (hunk 7 above)
Comment (NileshPant1999) [reply apps/travelperk/helpers.py:134]: this was for testing, uncommented

File: apps/travelperk/helpers.py (hunk 1 above)
Comment (NileshPant1999) [reply apps/travelperk/helpers.py:142]: @Shwetabhk i changed some code, here forget to change the comment, edited now

File: apps/travelperk/helpers.py (hunk 8 above)
Comment (NileshPant1999) [reply apps/travelperk/helpers.py:146]: changed the name

File: apps/travelperk/helpers.py (hunk 3 above)
Comment (NileshPant1999) [reply apps/travelperk/helpers.py:140]: yes @Shwetabhk we get empty array for data key if there are no transactions
the response looks something like this
{"count":0 , "offset": 0, "data": []}

**Review by Shwetabhk — APPROVED** [review]

**Pull request merged**